### PR TITLE
PowerBI major refinement in working of tool and tweaks in the rest

### PR DESCRIPTION
**Pull request by eavanvalkenburg**
# PowerBI major refinement in working of tool and tweaks in the rest

I've gained some experience with more complex sets and the earlier implementation had too many tries by the agent to create DAX, so refactored the code to run the LLM to create dax based on a question and then immediately run the same against the dataset, with retries and a prompt that includes the error for the retry. This works much better! 

Also did some other refactoring of the inner workings, making things clearer, more concise and faster.

@vowelparrot for review

### CODE REVIEW

File: langchain/tools/powerbi/tool.py [hunk 1]
```diff
@@ -23,110 +23,135 @@ class QueryPowerBITool(BaseTool):
...
+    ) -> LLMChain:
+        """Make sure the LLM chain has the correct input variables."""
+        if llm_chain.prompt.input_variables != [
+            "tool_input",
+            "tables",
+            "schemas",
+            "examples",
+        ]:
+            raise ValueError(
+                "LLM chain for InputToQueryTool must have input variables ['tool_input', 'tables', 'schemas', 'examples']"  # noqa: C0301 E501 # pylint: disable=C0301
```
Comment (dev2049) [inline langchain/tools/powerbi/tool.py:54]: would be nice to tell the user what variables it actually found

**Review by dev2049 — COMMENTED** [review]
one nit but otherwise lgtm

File: langchain/tools/powerbi/tool.py (hunk 1 above)
Comment (eavanvalkenburg) [reply langchain/tools/powerbi/tool.py]: good one, added

**Comment by dev2049** [comment]
@eavanvalkenburg ready to merge?

**Comment by eavanvalkenburg** [comment]
@dev2049 thanks for checking, give me another 30!

**Comment by eavanvalkenburg** [comment]
never mind, good to go, thanks @dev2049

**Pull request merged**